File: C++/072.cpp
// 072 Edit Distance
//
// 给定两个单词，word1 和 word2，转换 word1 到 word2，问最少要操作几次？还是同样的问题，动态规划，不过这个要难一丢丢。
//
// 规则给定，只允许你插入、删除和替换字符。
//
// 可参考这里：https://web.stanford.edu/class/cs124/lec/med.pdf

class Solution {
public:
    int minDistance(string word1, string word2) {
        int m = word1.size();
        int n = word2.size();
        vector<vector<int>> dp(m + 1, vector<int>(n + 1, 0));
        for (int i = 1; i <= m; i++) dp[i][0] = i;
        for (int j = 1; j <= n; j++) dp[0][j] = j;
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (word1[i - 1] == word2[j - 1]) {
                    dp[i][j] = dp[i - 1][j - 1];
                } else {
                    dp[i][j] = min(dp[i - 1][j - 1], min(dp[i][j - 1], dp[i - 1][j])) + 1;
                }
            }
        }
        return dp[m][n];
    }
};

File: Python/072.py
class Solution:
    def minDistance(self, w1: str, w2: str) -> int:
        m, n = len(w1), len(w2)
        dp = [0] * (n+1)
        dp[0] = 0
        for j in range(1, n+1):
            dp[j] = j
        for i in range(1, m+1):
            res = dp[0]
            dp[0] = i
            for j in range(1, n+1):
                tmp, res = res, dp[j]
                if w1[i-1] == w2[j-1]:
                    dp[j] = tmp
                else:
                    dp[j] = min(dp[j], dp[j-1], tmp)+1
        return dp[n]
